File: Application_part/project_final_version/app/src/main/java/com/example/heartbeat/BLEDataAdapter.java
package com.example.heartbeat;

import android.bluetooth.BluetoothDevice;
import android.companion.BluetoothLeDeviceFilter;
import android.content.Context;
import android.content.Intent;
import android.text.Layout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;


public class BLEDataAdapter extends RecyclerView.Adapter<ViewHolder>{

    public interface ItemClickListener{
        void onItemClick(View view, int position);
    }

    private ArrayList<BluetoothDevice> mBLEDevices = null;
    private ItemClickListener itemClickListener = null;



    BLEDataAdapter(ArrayList<BluetoothDevice> mBLEDevices){
        this.mBLEDevices = mBLEDevices;
    }
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        View view = inflater.inflate(R.layout.bledeviceitem, parent, false);
        ViewHolder viewHolder = new ViewHolder(view, itemClickListener);

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {

        final String BLEDeviceName = mBLEDevices.get(position).getName();
        final String BLEDeviceMAC = mBLEDevices.get(position).getAddress();

        if( BLEDeviceName != null && BLEDeviceName.length() > 0){
            viewHolder.DeviceTextView.setText(BLEDeviceName);
        }
        else{
            viewHolder.DeviceTextView.setText("기기 이름을 식별할 수 없습니다.");
        }
        viewHolder.DeviceMACTextView.setText(BLEDeviceMAC);


    }

    @Override
    public int getItemCount() {
        if(mBLEDevices != null) {
            return mBLEDevices.size();
        }
        else{
            return 0;
        }
    }

    public void addDevice(BluetoothDevice device){
        if(!mBLEDevices.contains(device)){
            mBLEDevices.add(device);
        }
    }

    public BluetoothDevice getDevice(int position){
        return mBLEDevices.get(position);
    }

    public void clear(){
        mBLEDevices.clear();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public void setItemClickListener(ItemClickListener listener){
        this.itemClickListener = listener;
    }


}

File: Application_part/project_kotlin_version/Kotlin_ver/app/src/main/java/com/example/kotlin_ver/BLEDataAdapter.kt
package com.example.kotlin_ver

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

class BLEDataAdapter(mBLEDevices : ArrayList<BluetoothDevice>) : RecyclerView.Adapter<ViewHolder>() {
    val mBLEDevices = mBLEDevices
    lateinit var itemClickListener : ItemClickListener

    interface ItemClickListener {
        fun onItemClick(view: View?, position: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = parent.context
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = inflater.inflate(R.layout.bledeviceitem, parent, false)

        return ViewHolder(view, itemClickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val BLEDeviceName = mBLEDevices[position].name
        val BLEDeviceMAC = mBLEDevices[position].address


        if (BLEDeviceName != null && BLEDeviceName.length > 0) {
            holder.deviceMACTextView.setText(BLEDeviceName)
        } else {
            holder.deviceTextView.setText("기기 이름을 식별할 수 없습니다.")
        }
        holder.deviceMACTextView.setText(BLEDeviceMAC)
        //holder.cardView.setOnClickListener(View.OnClickListener { println("터치터치함!") })
        holder.apply { bind() }
    }

    override fun getItemCount(): Int {
        if(mBLEDevices != null){
            return mBLEDevices.size
        }else{
            return 0
        }
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }


    fun addDevice(device : BluetoothDevice){
        if(!mBLEDevices.contains(device))mBLEDevices.add(device)
    }

    fun getDevice(position: Int) : BluetoothDevice{
        return mBLEDevices.get(position)
    }

    fun clearDevice(){
        mBLEDevices.clear()
    }



}